layers/+lang/c-c++/funcs.el

```diff
@@ -182,3 +185,161 @@ and the arguments for flyckeck-clang based on a project-specific text file."
+
+      (evil-set-initial-state '(spacemacs//c-c++-lsp-symbol nil "-tree-mode") 'emacs)
+      ;;evil-record-macro keybinding clobbers q in cquery-tree-mode-map for some reason?
+      (evil-make-overriding-map (symbol-value (spacemacs//c-c++-lsp-symbol nil "-tree-mode-map")))))
+
+(defun spacemacs//c-c++-lsp-setup-company ()
+  "Setup LSP backend auto-completion."
+    (progn
+      (spacemacs|add-company-backends :backends company-lsp :modes c-mode-common)
+      ;;Disable client-side cache and sorting, as server does a better job
+      (setq company-transformers nil company-lsp-async t company-lsp-cache-candidates nil)))
+
+(defun spacemacs//c-c++-lsp-wrap-functions ()
+  "Wrap navigation functions for the LSP backend specified by the `c-c++-backend' configuration variable."
+  (defun c-c++/call-hierarchy () (interactive) (spacemacs//c-c++-lsp-funcall-interactively nil "-call-hierarchy" nil))
+  (defun c-c++/call-hierarchy-inv () (interactive) (spacemacs//c-c++-lsp-funcall-interactively nil "-call-hierarchy" t))
+  (defun c-c++/inheritance-hierarchy () (interactive) (spacemacs//c-c++-lsp-funcall-interactively nil "-inheritance-hierarchy"))
+  (defun c-c++/inheritance-hierarchy-inv () (interactive) (spacemacs//c-c++-lsp-funcall-interactively nil "-inheritance-hierarchy" t))
+  (defun c-c++/member-hierarchy () (interactive) (spacemacs//c-c++-lsp-funcall-interactively-no-args nil "-member-hierarchy"))
+  (defun c-c++/freshen-index () (interactive) (spacemacs//c-c++-lsp-funcall-interactively nil "-freshen-index"))
+  (defun c-c++/preprocess-file () (interactive) (spacemacs//c-c++-lsp-funcall-interactively nil "-preprocess-file")))
+
+(defun spacemacs//c-c++-lsp-bind-keys-for-mode (mode)
+  "Bind LSP backend functions for the specificed mode."
+  (spacemacs/set-leader-keys-for-major-mode mode
+    ;; goto
+    ;; FIXME gf conflicts with `lsp-ui-flycheck-list' binding in lsp layer
+    ;; Better corrected in lsp-layer? rebinding here for now
+    "ge" #'lsp-ui-flycheck-list
+    "gf" 'find-file-at-point
+    "gF" 'ffap-other-window
+    "g&" #'c-c++/references-address
+    "gR" #'c-c++/references-read
+    "gW" #'c-c++/references-write
+    "gc" #'c-c++/callers
+    "gv" #'c-c++/vars
+    ;; help/hierarchy
+    "hb" #'c-c++/base ;;Replace this with lsp-goto-implementation in lsp-layer?
+    "hc" #'c-c++/call-hierarchy
+    "hC" #'c-c++/call-hierarchy-inv
+    "hi" #'c-c++/inheritance-hierarchy
+    "hI" #'c-c++/inheritance-hierarchy-inv
+    "hm" #'c-c++/member-hierarchy
+    ;; lsp/backend
+    "lf" #'c-c++/freshen-index
+    "lp" #'c-c++/preprocess-file))
+
+(defun spacemacs//c-c++-lsp-customise-lsp-ui-peek ()
+  "Wrap some backend-specific features using the `lsp-ui-peek' functions provided by the `lsp' layer (`lsp-ui' package)"
+  (spacemacs//c-c++-lsp-call-function "spacemacs//c-c++-lsp-" "-customise-lsp-ui-peek")
+
+  (defun c-c++/callers () (interactive) (lsp-ui-peek-find-custom 'callers (spacemacs//c-c++-lsp-string "$" "/callers")))
```

Comment:
In ccls, `c-c++/callers` should use 

`(lsp-ui-peek-find-custom 'base "$ccls/call")`

Reply:
Hey Fangrui,
Will make the changes Friday. A gentle suggestion from a lazy integrator... where it doesn't compromise your vision too much, keeping the ccls API close to that of cquery might encourage uptake, as it makes it easier to support both as drop-in alternatives.

Reply:
Thank you for your kind suggestion (and sorry for the hassle I caused I again). 

The hierarchy APIs are goodies I use a lot but don't expect much adoption by others. ccls is still young and not many people have noticed it. These stuff touched the area where I recently improved (flattened views of old hierarchy APIs) as I know someone favors them (ivy-xref helm-xref integration).

Reply:
No bother Fangrui. I've amended the commit - would you mind verifying the bindings on a known-good project when you get a chance? Both the callers (`, g c`) and base (`, h b`) bindings resulted in a language server timeout for me using ccls, but probably my project setup. Not actively working in C the last couple of weeks, so don't have time to investigate further at the moment.